sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/utils/avro/RowDeserializerCodeGenerator.java

```diff
@@ -0,0 +1,1180 @@
+import org.apache.beam.sdk.values.Row;
+import org.apache.commons.lang3.StringUtils;
+import org.codehaus.jackson.JsonNode;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RowDeserializerCodeGenerator<T> extends SerDesBase {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(RowDeserializerCodeGenerator.class);
+
+  private static final String DECODER = "decoder";
+  private static final String REUSE = "reuse";
+
+  private final Schema writer;
+  private final Schema reader;
+  private JDefinedClass deserializerClass;
+  private JFieldVar schemaMapField;
+  private JMethod generateBeamSchemasMethod;
+  private Map<Long, Schema> schemaMap = new HashMap<>();
+  private Map<Long, JVar> schemaVarMap = new HashMap<>();
+  private Map<String, JMethod> deserializeMethodMap = new HashMap<>();
+  private Map<String, JMethod> skipMethodMap = new HashMap<>();
+  private Map<JMethod, Set<Class<? extends Exception>>> exceptionFromMethodMap = new HashMap<>();
+  private JFieldVar avroSchemaMapField;
+
+
+  /**
+   * Row Deserializer Code generator class. Based on given schemas it generates Java code.
+   *
+   * @param writer           the writer's Avro schema
+   * @param reader           the reader's Avro schema
+   * @param destination      Path for generated java codes.
+   * @param classLoader      classLoader
+   * @param compileClassPath Path for compiled java classes.
+   */
+  public RowDeserializerCodeGenerator(Schema writer, Schema reader, File destination,
+      ClassLoader classLoader, String compileClassPath) {
+    super("deserializer", destination, classLoader, compileClassPath);
+    this.writer = writer;
+    this.reader = reader;
+    LOGGER.warn("RowDeserializerCodeGenerator Constructed.");
+  }
+
+  /**
+   * Row Deserializer Code generator. Based on given constructor's variable it generates Java Code
+   * and load it in classpath then return instance of the generated code.
+   *
+   * @return {@link RowDeserializer}
+   */
+  public RowDeserializer<T> generateDeserializer() {
+    LOGGER.warn("Start generating code of deserializer");
+    String className = getClassName(writer, reader, "RowDeserializer");
+    JPackage classPackage = codeModel._package(this.generatedPackageName);
+    LOGGER.warn("Start generating code of deserializer: " + className);
+
+    try {
+      deserializerClass = classPackage._class(className);
+
+      JVar readerSchemaVar = deserializerClass
+          .field(JMod.PRIVATE | JMod.FINAL, Schema.class, "avroSchema");
+      JMethod constructor = deserializerClass.constructor(JMod.PUBLIC);
+      JVar constructorParam = constructor.param(Schema.class, "avroSchema");
+      constructor.body().assign(JExpr.refthis(readerSchemaVar.name()), constructorParam);
+
+      Schema aliasedWriterSchema = Schema.applyAliases(writer, reader);
+      Symbol resolvingGrammar = new ResolvingGrammarGenerator()
+          .generate(aliasedWriterSchema, reader);
```

Comment:
I would use [Resolver.resolve](https://avro.apache.org/docs/current/api/java/org/apache/avro/Resolver.html#resolve-org.apache.avro.Schema-org.apache.avro.Schema-) here rather than attempting to consume the grammar directly.  It was made for basically this purpose, building your own decoders, but without having to understand the underlying complexities of the grammar.  See also [refactoring-resolution](https://github.com/apache/avro/blob/master/doc-deprecated/src/content/mddocs/refactoring-resolution.md).  imo using the resolver is significantly easier and less error prone than trying to interpret the grammar directly.